Core/Src/main.c

```diff
@@ -86,18 +112,60 @@ int main(void)
 
   /* Initialize all configured peripherals */
   MX_GPIO_Init();
+  MX_DMA_Init();
   MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM1_Init();
   /* USER CODE BEGIN 2 */
 
+
+
+  // Enable Chip Select on default
+  HAL_GPIOWritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_SET);
+
+
+  // Serial terminal test
+  uart_buffer_length = sprintf(uart_buffer, "Serial terminal test\r\n");
+  HAL_UART_Transmit(&huart2, (uint8_t *) uart_buffer, uart_buffer_length, 100);
+
+
+  //Start the timer for PWM
+  HAL_TIM_PWM_Start(&htim1, TIM_CHANNEL_1);
+
+
+
   /* USER CODE END 2 */
 
   /* Infinite loop */
   /* USER CODE BEGIN WHILE */
   while (1)
   {
+
+	// One SPI communication cycle with MCP3004 ADC
+	HAL_GPIOWritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_RESET);
+	HAL_SPI_TransmitRecieve(&hspi1, spi_output_buffer, spi_input_buffer, 3, 100);
+	HAL_GPIOWritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_SET);
+
+	//Extract 10 bit data in spi_input_buffer to spi_input_value
+	spi_input_value = (((uint16_t)(spi_input_buffer[1] & 0b00000011)) << 8)
+			          + (uint16_t) spi_input_buffer[2];
```

Comment:
once again try using hex... once you get used to the number system and understand that 1 hex number represents 4 binary numbers, it becomes very easy to go between the two.

Reply:
Bit manips look good tho!